src/adapter/src/catalog.rs

```diff
@@ -2611,6 +2616,69 @@ impl<S: Append> Catalog<S> {
         Ok(location)
     }
 
+    pub fn resolve_storage_instance(
+        &self,
+        storage_instance_config: StorageInstanceConfig,
+    ) -> Result<StorageInstanceSizeOrAddress, AdapterError> {
+        let instance_sizes = &self.state.cluster_replica_sizes;
+        let instance_size_or_address = match storage_instance_config {
+            StorageInstanceConfig::Remote { addr } => StorageInstanceSizeOrAddress::Remote { addr },
+            StorageInstanceConfig::Managed { size } => {
+                let allocation = instance_sizes.0.get(&size).ok_or_else(|| {
+                    let mut entries = instance_sizes.0.iter().collect::<Vec<_>>();
+                    entries.sort_by_key(
+                        |(
+                            _name,
+                            ComputeInstanceReplicaAllocation {
+                                scale, cpu_limit, ..
+                            },
+                        )| (scale, cpu_limit),
+                    );
+                    let expected = entries
+                        .into_iter()
+                        .filter(|(_, allocation)| allocation.scale == NonZeroUsize::new(1).unwrap())
+                        .map(|(name, _)| name.clone())
+                        .collect();
+                    AdapterError::InvalidStorageInstanceSize {
+                        size: size.clone(),
+                        expected,
+                    }
+                })?;
+
+                // Forbid sizes with multiple instances
+                if allocation.scale > NonZeroUsize::new(1).unwrap() {
+                    return Err(AdapterError::InvalidStorageInstanceScale { size: size.clone() });
+                }
+                StorageInstanceSizeOrAddress::Managed {
+                    allocation: StorageInstanceResourceAllocation {
+                        memory_limit: allocation.memory_limit,
+                        cpu_limit: allocation.cpu_limit,
+                        workers: allocation.workers,
+                    },
+                    size,
+                }
+            }
+            StorageInstanceConfig::Undefined => {
+                // When no size specifier is given, the smallest instance size available is chosen instead
+                let (size, allocation) = instance_sizes
+                    .0
+                    .iter()
+                    .filter(|(_, a)| a.scale == NonZeroUsize::new(1).unwrap())
+                    .min_by_key(|(_, a)| a.memory_limit)
+                    .expect("should have at least one valid storage instance size");
+                StorageInstanceSizeOrAddress::Managed {
+                    allocation: StorageInstanceResourceAllocation {
+                        memory_limit: allocation.memory_limit,
+                        cpu_limit: allocation.cpu_limit,
+                        workers: allocation.workers,
+                    },
+                    size: size.clone(),
+                }
```

Comment:
Ideally this would be controlled by a `--default-storage-host-size` parameter to environmentd so the cloud team can dial around the default size as they like!

Reply:
Ah, I was following the design doc in defaulting to the smallest one available! Lemme add the option to have a fixed default!

Reply:
Implemented in 20e07793e71d0ec416b99ebc7fbf347b3755574c!

Reply:
Ah sorry, never did a final pass on that design doc. Having explicit control over these defaults is a philosophical shift we made since I originally wrote the design doc. 🙈